content/articles/lru-cache-implementation-in-javascript/index.md

```diff
@@ -0,0 +1,192 @@
+
+From the example above, we have performed many operations while building the LRU cache. These operations include:
+
+#### 1. Searching
+When we want to get the color `red` from our list, first, we need to check our entire list to check whether the item is there. 
+
+If the element is present, we reorder the list and move `red` to the beginning of the list. In this case, the time complexity for the re-ordering will be `O(n)`.
+
+Time complexity is the amount of time a block of code takes to execute. I recommend [this time complexity](https://www.mygreatlearning.com/blog/why-is-time-complexity-essential/) article, to learn about time complexity.
+
+#### 2. Adding items
+When adding an item to the list, the time complexity of the operation is O(n).
+
+We can optimize the shifting operation by eliminating it. To do so we use a double liked list. How?
+
+Still using the example `[red, green, white, blue]` we needed to add `black`. Since it wasn't in our list,we removed the oldest element `blue` and shifted the entire list of items.
+
+In a doubly-linked list, all you need to do is to remove the rear item and update the rear element to point to the previous element. Then add a new node and make it point to the second element and update the front. The entire operation takes 0(1) time complexity.
+
+This will be clearly seen in the code example below.
+
+We can also optimize searching. How?
+
+Searching can be optimized by using a hash table. With a Hash, you will have a key and a value pair. The key in this case will be the actual values in our cache. The value will be the address of our value as shown in the diagram below.
+
+![Hash Table & Double Linked List Diagram](/ddl_hash.png)
+
+To add `black` to our list, by using a doubly-linked list, we will remove the rear element `blue` from our list. `Blue` in this case is our least recently accessed item. Then, update the rear element to point to the previous element `white`. 
+
+In our hash table, we will remove the value(address) that is associated with our key `blue`. Therefore, the value for our key `blue` will be null in the hashtable. A new node is created where we put `black`. The previous front gets updated to point to the node that has `black`. `Black` will also have an address associated with it as well.
+
+The shifting, in this case, is done in 0(1) time complexity.
+
+Before adding `black`, we need to search first and see if it is available in the cache. To search for black, we go to the hash table. In the Hash table, key `black` will not have a corresponding value attached to it. This means that it is not present in our cache. This allows us to find out whether an item is in the cache with just O(1) time complexity.
+
+### Implementing LRU Cache in JavaScript.
+
+- Visit [leetecode](https://leetcode.com/)
+- Sign in to the account
+- Visit [lru-cache](https://leetcode.com/problems/lru-cache/)
+- Go through the problem statement.
+- Implement the lrucache class.

```

Comment:
Provide instructions on how to run the code in this article without Signing up at LeetCode. Use the browser console to guide the reader on running the code. Only leave a link to the LeetCode question.